{"cells":[{"cell_type":"markdown","metadata":{"id":"GNBQC47DjGQ-"},"source":["### BIC3007 - *NGS Data Processing and Variant Calling*"]},{"cell_type":"markdown","metadata":{"id":"-oKLRtT7jGRF"},"source":["# Analisis Variasi Genetik Menggunakan Data NGS"]},{"cell_type":"markdown","metadata":{"id":"uiW4kkT7jGRG"},"source":["## Pengantar"]},{"cell_type":"markdown","metadata":{"id":"m0_Jcu7jjGRH"},"source":["Pada tutorial ini, kita akan melakukan pemrosesan data NGS (*short-read Amplicon Sequencing*) and melakukan pemanggilan variasi. Sampel DNA berasa dari ekstraksi pada jaringan darah manusia. DNA hasil ekstraksi kemudian dikirim untuk dilakukan NGS. Detail untuk pemrosesan sampel dapat dibaca pada publikasi [berikut](https://www.nature.com/articles/gim2016218#Sec2), tepatnya bagian *PCR amplification, pooling, library construction, and next-generation sequencing*."]},{"cell_type":"markdown","metadata":{"id":"3e3Qxt4fjGRI"},"source":["Kita akan melakukan analisis pada gen HBA2 yang berasa dari dua sampel. Tutorial ini bertujuan untuk memperkenalkan alur pemrosesan data NGS dan *variant calling* dan *softwares/tools* yang dapat digunakan. Untuk detail terkait fungsi masing-masing langkah dan program yang digunakan, mohon untuk mempelajari secara mandiri."]},{"cell_type":"markdown","metadata":{"id":"nR56NBrtjGRI"},"source":["Secara singkat, berikut alur kerja akan dilakukan\n","1. Buat indeks genom referensi untuk digunakan oleh bwa dan samtools\n","2. Sejajarkan sekuen dengan genom referensi\n","3. Ubah format pensejajaran menjadi *sorted BAM*\n","4. Hitung *read coverage* posisi pada genom\n","5. Deteksi polimorfisme nukleotida tunggal (SNP)\n","6. Filter dan laporkan varian SNP dalam VCF (variant calling format) \n","\n","Tutorial ini dibuat berdasarkan script yang dibuat oleh [Jason J Willian](https://jasonjwilliamsny.github.io/wrangling-genomics/02-variant-calling-workflow.html) dengan modifikasi. *Software/tools* yang kita gunakan disini tidak berada di lingkungan melainkan lingkungan terminal, sehingga disini kita menggunakan *shell scripting*. Untuk setiap command harus di *run* (*copy/paste*) ke terminal, atau dapat pula di run disini namun dengan sebelumnya menambahkan character \"!\" tepat sebelum setiap command."]},{"cell_type":"markdown","metadata":{"id":"bFZDBU9ljGRO"},"source":["## Installation"]},{"cell_type":"markdown","metadata":{"id":"z0u6set6jGRO"},"source":["Sebelumnya kita perlu menginstal *softwares/tools* yang akan digunakan. *Copy/paste* command di bawah ke terminal, satu per satu. "]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"YmaxXGgijGRP","executionInfo":{"status":"ok","timestamp":1652768307602,"user_tz":-480,"elapsed":281226,"user":{"displayName":"YUSTIKA SARI","userId":"17498490950316481851"}},"outputId":"021014ce-d4ba-4aad-f9bf-02da92d3510d"},"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting package metadata (current_repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n","Solving environment: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n","\n","## Package Plan ##\n","\n","  environment location: /usr/local\n","\n","  added / updated specs:\n","    - bwa\n","\n","\n","The following packages will be downloaded:\n","\n","    package                    |            build\n","    ---------------------------|-----------------\n","    bwa-0.7.17                 |       h7132678_9         187 KB  bioconda\n","    ca-certificates-2021.10.8  |       ha878542_0         139 KB  conda-forge\n","    certifi-2021.10.8          |   py37h89c1867_2         145 KB  conda-forge\n","    conda-4.12.0               |   py37h89c1867_0         1.0 MB  conda-forge\n","    libnsl-2.0.0               |       h7f98852_0          31 KB  conda-forge\n","    openssl-1.1.1o             |       h166bdaf_0         2.1 MB  conda-forge\n","    perl-5.32.1                | 2_h7f98852_perl5        14.4 MB  conda-forge\n","    ------------------------------------------------------------\n","                                           Total:        18.0 MB\n","\n","The following NEW packages will be INSTALLED:\n","\n","  bwa                bioconda/linux-64::bwa-0.7.17-h7132678_9\n","  libnsl             conda-forge/linux-64::libnsl-2.0.0-h7f98852_0\n","  perl               conda-forge/linux-64::perl-5.32.1-2_h7f98852_perl5\n","\n","The following packages will be UPDATED:\n","\n","  ca-certificates                      2020.12.5-ha878542_0 --> 2021.10.8-ha878542_0\n","  certifi                          2020.12.5-py37h89c1867_1 --> 2021.10.8-py37h89c1867_2\n","  conda                                4.9.2-py37h89c1867_0 --> 4.12.0-py37h89c1867_0\n","  openssl                                 1.1.1j-h7f98852_0 --> 1.1.1o-h166bdaf_0\n","\n","\n","\n","Downloading and Extracting Packages\n","conda-4.12.0         | 1.0 MB    | : 100% 1.0/1 [00:00<00:00,  2.43it/s]\n","libnsl-2.0.0         | 31 KB     | : 100% 1.0/1 [00:00<00:00, 29.06it/s]\n","perl-5.32.1          | 14.4 MB   | : 100% 1.0/1 [00:05<00:00,  5.04s/it]\n","bwa-0.7.17           | 187 KB    | : 100% 1.0/1 [00:00<00:00, 16.31it/s]\n","openssl-1.1.1o       | 2.1 MB    | : 100% 1.0/1 [00:00<00:00,  2.14it/s]\n","ca-certificates-2021 | 139 KB    | : 100% 1.0/1 [00:00<00:00, 19.51it/s]\n","certifi-2021.10.8    | 145 KB    | : 100% 1.0/1 [00:00<00:00,  2.96it/s]\n","Preparing transaction: - \b\b\\ \b\bdone\n","Verifying transaction: / \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n","Executing transaction: | \b\b/ \b\b- \b\bdone\n","Collecting package metadata (current_repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n","Solving environment: / \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n","\n","## Package Plan ##\n","\n","  environment location: /usr/local\n","\n","  added / updated specs:\n","    - samtools\n","\n","\n","The following packages will be downloaded:\n","\n","    package                    |            build\n","    ---------------------------|-----------------\n","    samtools-1.3.1             |                0         1.5 MB  bioconda\n","    ------------------------------------------------------------\n","                                           Total:         1.5 MB\n","\n","The following NEW packages will be INSTALLED:\n","\n","  samtools           bioconda/linux-64::samtools-1.3.1-0\n","\n","\n","\n","Downloading and Extracting Packages\n","samtools-1.3.1       | 1.5 MB    | : 100% 1.0/1 [00:00<00:00,  2.47it/s]\n","Preparing transaction: / \b\bdone\n","Verifying transaction: \\ \b\bdone\n","Executing transaction: / \b\bdone\n","Collecting package metadata (current_repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n","Solving environment: / \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n","\n","## Package Plan ##\n","\n","  environment location: /usr/local\n","\n","  added / updated specs:\n","    - bcftools\n","\n","\n","The following packages will be downloaded:\n","\n","    package                    |            build\n","    ---------------------------|-----------------\n","    bcftools-1.15.1            |       h0ea216a_0         863 KB  bioconda\n","    c-ares-1.18.1              |       h7f98852_0         113 KB  conda-forge\n","    cudatoolkit-11.1.1         |      h6406543_10        1.20 GB  conda-forge\n","    gsl-2.7                    |       he838d99_0         3.2 MB  conda-forge\n","    htslib-1.15.1              |       h9753748_0         2.1 MB  bioconda\n","    keyutils-1.6.1             |       h166bdaf_0         115 KB  conda-forge\n","    krb5-1.19.3                |       h3790be6_0         1.4 MB  conda-forge\n","    libblas-3.9.0              |   12_linux64_mkl          12 KB  conda-forge\n","    libcblas-3.9.0             |   12_linux64_mkl          12 KB  conda-forge\n","    libcurl-7.83.1             |       h7bff187_0         342 KB  conda-forge\n","    libdeflate-1.10            |       h7f98852_0          77 KB  conda-forge\n","    libnghttp2-1.47.0          |       h727a467_0         808 KB  conda-forge\n","    libssh2-1.10.0             |       ha56f1ee_2         233 KB  conda-forge\n","    libstdcxx-ng-12.1.0        |      ha89aaad_16         4.3 MB  conda-forge\n","    python_abi-3.7             |          2_cp37m           4 KB  conda-forge\n","    ------------------------------------------------------------\n","                                           Total:        1.22 GB\n","\n","The following NEW packages will be INSTALLED:\n","\n","  bcftools           bioconda/linux-64::bcftools-1.15.1-h0ea216a_0\n","  cudatoolkit        conda-forge/linux-64::cudatoolkit-11.1.1-h6406543_10\n","  gsl                conda-forge/linux-64::gsl-2.7-he838d99_0\n","  htslib             bioconda/linux-64::htslib-1.15.1-h9753748_0\n","  keyutils           conda-forge/linux-64::keyutils-1.6.1-h166bdaf_0\n","  libblas            conda-forge/linux-64::libblas-3.9.0-12_linux64_mkl\n","  libcblas           conda-forge/linux-64::libcblas-3.9.0-12_linux64_mkl\n","  libdeflate         conda-forge/linux-64::libdeflate-1.10-h7f98852_0\n","\n","The following packages will be UPDATED:\n","\n","  c-ares                                  1.17.1-h7f98852_1 --> 1.18.1-h7f98852_0\n","  krb5                                    1.17.2-h926e7f8_0 --> 1.19.3-h3790be6_0\n","  libcurl                                 7.75.0-hc4aaa36_0 --> 7.83.1-h7bff187_0\n","  libnghttp2                              1.43.0-h812cca2_0 --> 1.47.0-h727a467_0\n","  libssh2                                  1.9.0-ha56f1ee_6 --> 1.10.0-ha56f1ee_2\n","  libstdcxx-ng                            9.3.0-h6de172a_18 --> 12.1.0-ha89aaad_16\n","  python_abi                                    3.7-1_cp37m --> 3.7-2_cp37m\n","\n","\n","\n","Downloading and Extracting Packages\n","c-ares-1.18.1        | 113 KB    | : 100% 1.0/1 [00:00<00:00, 10.90it/s]\n","libstdcxx-ng-12.1.0  | 4.3 MB    | : 100% 1.0/1 [00:00<00:00,  1.21it/s]\n","gsl-2.7              | 3.2 MB    | : 100% 1.0/1 [00:00<00:00,  1.16it/s]\n","libdeflate-1.10      | 77 KB     | : 100% 1.0/1 [00:00<00:00, 19.96it/s]\n","htslib-1.15.1        | 2.1 MB    | : 100% 1.0/1 [00:00<00:00,  2.13it/s]\n","krb5-1.19.3          | 1.4 MB    | : 100% 1.0/1 [00:00<00:00,  2.37it/s]\n","libssh2-1.10.0       | 233 KB    | : 100% 1.0/1 [00:00<00:00,  8.99it/s]\n","libcblas-3.9.0       | 12 KB     | : 100% 1.0/1 [00:00<00:00, 27.67it/s]\n","python_abi-3.7       | 4 KB      | : 100% 1.0/1 [00:00<00:00, 18.84it/s]\n","keyutils-1.6.1       | 115 KB    | : 100% 1.0/1 [00:00<00:00, 15.08it/s]\n","libblas-3.9.0        | 12 KB     | : 100% 1.0/1 [00:00<00:00, 26.67it/s]\n","bcftools-1.15.1      | 863 KB    | : 100% 1.0/1 [00:00<00:00,  4.04it/s]\n","libcurl-7.83.1       | 342 KB    | : 100% 1.0/1 [00:00<00:00,  7.63it/s]\n","cudatoolkit-11.1.1   | 1.20 GB   | : 100% 1.0/1 [03:39<00:00, 219.17s/it]              \n","libnghttp2-1.47.0    | 808 KB    | : 100% 1.0/1 [00:00<00:00,  4.43it/s]\n","Preparing transaction: \\ \b\bdone\n","Verifying transaction: / \b\b- \b\b\\ \b\b| \b\bdone\n","Executing transaction: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ By downloading and using the CUDA Toolkit conda packages, you accept the terms and conditions of the CUDA End User License Agreement (EULA): https://docs.nvidia.com/cuda/eula/index.html\n","\n","\b\bdone\n"]}],"source":["conda install -c bioconda bwa #install bwa dari biocaonda\n","conda install -c bioconda samtools #install samtools\n","conda install -c bioconda bcftools #install bcftools"]},{"cell_type":"markdown","metadata":{"id":"-px6WRsWjGRQ"},"source":["Kita juga akan menggunakan *Integrative Genomics Viewers* (IGV) untuk membantu visualisasi hasil pemetaan dan/atau pensejajaran. Pada tutorial ini kita akan menggunakan IGV versi website nya, yang bisa diakses pada laman (https://igv.org/app/) atau klik [laman](https://igv.org/app/)."]},{"cell_type":"markdown","metadata":{"id":"BxQlECO4jGRQ"},"source":["Selanjutnya, kita akan buat *directory* untuk setiap file output untuk mempermudah akses. "]},{"cell_type":"code","execution_count":null,"metadata":{"id":"YyTLlNzejGRR"},"outputs":[],"source":["! mkdir -p results/sai\n","! mkdir -p results/sam\n","! mkdir -p results/bam\n","! mkdir -p results/bcf\n","! mkdir -p results/vcf\n","! mkdir -p data/ref_genome #folder untuk genom referensi"]},{"cell_type":"markdown","metadata":{"id":"JSbvAvHSjGRR"},"source":["## Download Genom Referensi"]},{"cell_type":"markdown","metadata":{"id":"6gesuVnrjGRR"},"source":["Referensi Genom manusia dapat diunduh pada laman [*Genome Browser Gateway*](https://genome-asia.ucsc.edu/cgi-bin/hgGateway?redirect=manual&source=genome.ucsc.edu) dari *Genomics Instutite UCSC*. Kita akan mengunduh sekuen beserta anotasi nya menggunakan http. Pada tutorial ini, kita hanya akan mengunduh sekuen genom kromosom 16 (lokasi gen HBA). Kita bisa mengunduh secara langsung pada [laman](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/chromosomes/) (pilih chr16) dan memindahkan secara manual ke folder data/ref_genome, atau dengan menggunakan command dibawah. "]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"355zQPBPjGRS","executionInfo":{"status":"ok","timestamp":1652768373246,"user_tz":-480,"elapsed":2234,"user":{"displayName":"YUSTIKA SARI","userId":"17498490950316481851"}},"outputId":"46528109-2af0-4b66-909c-b4f1e259f986"},"outputs":[{"output_type":"stream","name":"stdout","text":["--2022-05-17 06:17:02--  http://hgdownload.soe.ucsc.edu/goldenPath/hg19/chromosomes/chr16.fa.gz\n","Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163\n","Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:80... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 25667827 (24M) [application/x-gzip]\n","Saving to: ‘chr16.fa.gz’\n","\n","chr16.fa.gz         100%[===================>]  24.48M  17.6MB/s    in 1.4s    \n","\n","2022-05-17 06:17:04 (17.6 MB/s) - ‘chr16.fa.gz’ saved [25667827/25667827]\n","\n"]}],"source":["! wget http://hgdownload.soe.ucsc.edu/goldenPath/hg19/chromosomes/chr16.fa.gz #unduh Ref Genome chr16\n","! mv chr16.fa.gz data/ref_genome/ #pindah ref genom ke folder data"]},{"cell_type":"markdown","metadata":{"id":"5RJ7QF50jGRS"},"source":["## Indeksasi Genom Referensi "]},{"cell_type":"markdown","metadata":{"id":"dnFHrnIcjGRS"},"source":["Anda tidak dapat melakukan pemetaan secara langsung pada file .fasta, sehingga kita perlu mengindeks file genom referensi terlebih dahulu. Kita melakukan indeksasi untuk bwa dan samtools. Secara sederhana, *genome indexing* dapat artikan sebagai suatu bentuk pemrosesan untuk memampatkan ukuran dari text/informasi sehingga mempercepat suatu pencarian, dalam hal ini pemetaan sekuen. "]},{"cell_type":"markdown","metadata":{"id":"AWJSyDpsjGRT"},"source":["Command indeksasi genom dengan bwa. Silahkan baca detail command bwa [disini](http://bio-bwa.sourceforge.net/bwa.shtml)."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Yb3rnVATjGRT","executionInfo":{"status":"ok","timestamp":1652768642559,"user_tz":-480,"elapsed":148113,"user":{"displayName":"YUSTIKA SARI","userId":"17498490950316481851"}},"outputId":"b715b21b-c26f-4b23-b823-d72365816d0f"},"outputs":[{"output_type":"stream","name":"stdout","text":["[bwa_index] Pack FASTA... 1.50 sec\n","[bwa_index] Construct BWT for the packed sequence...\n","[BWTIncCreate] textLength=180709506, availableWord=24715364\n","[BWTIncConstructFromPacked] 10 iterations done. 40768818 characters processed.\n","[BWTIncConstructFromPacked] 20 iterations done. 75316370 characters processed.\n","[BWTIncConstructFromPacked] 30 iterations done. 106018306 characters processed.\n","[BWTIncConstructFromPacked] 40 iterations done. 133302226 characters processed.\n","[BWTIncConstructFromPacked] 50 iterations done. 157548210 characters processed.\n","[BWTIncConstructFromPacked] 60 iterations done. 179094066 characters processed.\n","[bwt_gen] Finished constructing BWT in 61 iterations.\n","[bwa_index] 109.66 seconds elapse.\n","[bwa_index] Update BWT... 0.62 sec\n","[bwa_index] Pack forward-only FASTA... 1.06 sec\n","[bwa_index] Construct SA from BWT and Occ... 33.03 sec\n","[main] Version: 0.7.17-r1188\n","[main] CMD: bwa index -p chr16bwaidx -a bwtsw data/ref_genome/chr16.fa.gz\n","[main] Real time: 147.501 sec; CPU: 145.893 sec\n"]}],"source":["! bwa index -p chr16bwaidx -a bwtsw data/ref_genome/chr16.fa.gz #indeksasi genom\n","! mv chr16* data/ref_genome #pindah file ke folder reference"]},{"cell_type":"markdown","metadata":{"id":"gxDvGrQ2jGRU"},"source":["Command indeksasi genom dengan bwa. Silahkan baca detail command samtools [disini](http://www.htslib.org/doc/samtools.html)."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"MEimsiZEjGRU"},"outputs":[],"source":["! gzip -d  data/ref_genome/chr16.fa.gz #unzip .fa.gz menjadi .fa\n","! samtools faidx data/ref_genome/chr16.fa #buat indeks (.fa.fai)"]},{"cell_type":"markdown","metadata":{"id":"WPn_nuatjGRU"},"source":["## Pensejajaran sekuen dengan genom referensi"]},{"cell_type":"markdown","metadata":{"id":"2lvjaiYwjGRV"},"source":["Kita lakukan pensejajaran bacaan-pendek (*short-reads*) ke genom referensi. Disini kita melakukan pemetaan sekuen forward/reverse ke genom referensi yang sudah kita indeks. Kita akan menghasilkan output dengan format SAI yang selanjutnya akan kita gabung convert bersama *paired-end* nya (forward dan reverse nya) menjadi format SAM. "]},{"cell_type":"markdown","metadata":{"id":"zPp4a6sIjGRV"},"source":["Command pensejajaran untuk sekuen dari masing-masing primer.  "]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"MxI0Xd1njGRV","executionInfo":{"status":"ok","timestamp":1652768483737,"user_tz":-480,"elapsed":951,"user":{"displayName":"YUSTIKA SARI","userId":"17498490950316481851"}},"outputId":"38f22772-4f5b-4c01-b188-48d5b00191b9"},"outputs":[{"output_type":"stream","name":"stdout","text":["[bwa_aln] 17bp reads: max_diff = 2\n","[bwa_aln] 38bp reads: max_diff = 3\n","[bwa_aln] 64bp reads: max_diff = 4\n","[bwa_aln] 93bp reads: max_diff = 5\n","[bwa_aln] 124bp reads: max_diff = 6\n","[bwa_aln] 157bp reads: max_diff = 7\n","[bwa_aln] 190bp reads: max_diff = 8\n","[bwa_aln] 225bp reads: max_diff = 9\n","[bwa_aln] fail to locate the index\n","[bwa_aln] 17bp reads: max_diff = 2\n","[bwa_aln] 38bp reads: max_diff = 3\n","[bwa_aln] 64bp reads: max_diff = 4\n","[bwa_aln] 93bp reads: max_diff = 5\n","[bwa_aln] 124bp reads: max_diff = 6\n","[bwa_aln] 157bp reads: max_diff = 7\n","[bwa_aln] 190bp reads: max_diff = 8\n","[bwa_aln] 225bp reads: max_diff = 9\n","[bwa_aln] fail to locate the index\n"]}],"source":["#Sampel 1\n","! bwa aln -t 4 data/ref_genome/chr16bwaidx data/trimmed_fastq/1_HBA2_1.fq.gz > 1_HBA2_1.sai #align bacaan forward primer dri sampel 1\n","! bwa aln -t 4 data/ref_genome/chr16bwaidx data/trimmed_fastq/1_HBA2_2.fq.gz > 1_HBA2_2.sai #align bacaan reverse primer dri sampel 1\n","! mv *.sai results/sai/ #pindah output .sai ke folder result"]},{"cell_type":"markdown","metadata":{"id":"OBA0rD7qjGRW"},"source":["Silahkan tulis command untuk sampel yang kedua, "]},{"cell_type":"code","execution_count":null,"metadata":{"id":"CzDdFQPrjGRW"},"outputs":[],"source":["#Sampel 2\n"]},{"cell_type":"markdown","metadata":{"id":"a1z8pIQ1jGRX"},"source":["Selanjutnya, kita gabungkan pembacaan forward dan reverse (*paired-end*) lalu kita konversi ke format SAM"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Mk2DqtMIjGRX","executionInfo":{"status":"ok","timestamp":1652768483739,"user_tz":-480,"elapsed":15,"user":{"displayName":"YUSTIKA SARI","userId":"17498490950316481851"}},"outputId":"e6abc9dd-fe63-401c-f1dd-6132ad4177b1"},"outputs":[{"output_type":"stream","name":"stdout","text":["[bwa_sai2sam_pe] fail to locate the index\n"]}],"source":["#Sampel 1\n","! bwa sampe data/ref_genome/chr16bwaidx results/sai/1_HBA2_1.sai results/sai/1_HBA2_2.sai data/trimmed_fastq/1_HBA2_1.fq.gz data/trimmed_fastq/1_HBA2_2.fq.gz > 1_HBA2.sam #konversi ke SAM\n","! mv *.sam results/sam #pindah output .sam ke folder result"]},{"cell_type":"markdown","metadata":{"id":"7RC4MMwgjGRY"},"source":["Silahkan tulis command untuk sampel yang kedua, "]},{"cell_type":"code","execution_count":null,"metadata":{"id":"XaESd1wMjGRY"},"outputs":[],"source":["#Sampel 2\n"]},{"cell_type":"markdown","metadata":{"id":"R-XwAjvGjGRY"},"source":["Note: Sequence Alignment Map (SAM) adalah format berbasis text yang digunakan untuk menyimpan pensejajaran sekuen biologi ke sekuen referensinya. Format SAM dan samtools dikembangkan [Heng Li and Bob Handsaker et al.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2723002/) Isi dari file SAM dapat langsung dilihat dibaca oleh manusia namun tidak sulit untuk dibaca dan diproses komputer. Untuk memungkinkan pemrosesan sekuen dengan program komputer kita perlu menkonversi SAM file ke BAM file. \n","\n","Lihat sebagian output file SAM:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"7UBU08kTjGRZ"},"outputs":[],"source":["! cat results/sam/1_HBA2.sam | head"]},{"cell_type":"markdown","metadata":{"id":"_JXAbnhmjGRZ"},"source":["## Konversi SAM ke (sorted) BAM"]},{"cell_type":"markdown","metadata":{"id":"QpPupN73jGRZ"},"source":["Binary Alignment Map (BAM) adalah representasi format SAM dalam bentuk binary yang terkompresi. File dengan format ini, tidak dapat dibaca manusia, untuk dapat membaca isinya kita perlu membukanya dengan software/tools. Format BAM diperlukan untuk analisis dengan software/tools yang umumnya hanya menerima input berupa text [binary](https://en.wikipedia.org/wiki/Binary_code). setelah konversi, kita juga akan melakukan *sorting* dan *indexing*. BAI file diperlukan untuk visualisasi dengan IGV."]},{"cell_type":"markdown","metadata":{"id":"G7jItNHzjGRZ"},"source":["Command untuk konversi ke BAM, *sorting* BAM dan buat indeks BAM (BAI)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"XfMidck7jGRZ"},"outputs":[],"source":["#Sampel 1\n","! samtools view -bS results/sam/1_HBA2.sam > 1_HBA2.bam #konversi SAM ke BAM\n","! samtools sort 1_HBA2.bam -o 1_HBA2_sorted.bam #sort BAM file\n","! samtools index 1_HBA2_sorted.bam #buat indeks BAM \n","! mv *.bam *.bai results/bam #pindah output .bam dan .bai ke folder result"]},{"cell_type":"markdown","metadata":{"id":"aou1j0B7jGRa"},"source":["Silahkan tulis command untuk sampel yang kedua, "]},{"cell_type":"code","execution_count":null,"metadata":{"id":"L2a9Q-AyjGRa"},"outputs":[],"source":["#Sampel 2\n"]},{"cell_type":"markdown","metadata":{"id":"fF-v4T60jGRa"},"source":["### Visualisasi pemetaan dengan IGV"]},{"cell_type":"markdown","metadata":{"id":"dKRcinE5jGRa"},"source":["Sampai, sini kita sudah mendapatkan hasil pemetaan sekuen-pendek (*sort-reads*) ke genom referensi. Kita bisa melihat secara langsung visualisasi pemetaannya melalui IGV. Silahkan akses [laman](https://igv.org/app/) dan ikuti step berikut."]},{"cell_type":"markdown","metadata":{"id":"8pv4UBxKjGRb"},"source":["1. Load file \n","\n","Tracks ==> Local File ... (silahkan pilih file *sorted.bam* dan *sorted.bam.bai*)\n","\n","2. Atur referensi\n","\n","Ganti lokasi kromosom dari *all* menjadi chr16, pada box di samping chr16 silahkan tulis **chr16:222,500-228,000** untuk menunjukkan tepat lokasi gen HBA\n","\n","Kita bisa menyesuaikan ukuran tampilan dengan mengubah menu bar pada kanan atas. "]},{"cell_type":"markdown","metadata":{"id":"INVlYmc3jGRb"},"source":["![Screenshot from 2021-04-28 23-58-26.png](attachment:dc1587e5-2c53-4d20-a7b6-56c8f93a3620.png)"]},{"cell_type":"markdown","metadata":{"id":"kDbsxSd3jGRb"},"source":["## Menghitung *read coverage* posisi pada genom"]},{"cell_type":"markdown","metadata":{"id":"fV7qnogtjGRb"},"source":["BCFtools adalah program untuk memanggil dan memanipulasi file dalam Variant Call Format (VCF) dan bentuk binary nya (BCF). Silahkan baca detail mengenai bcftools [disini](http://samtools.github.io/bcftools/bcftools.html). Untuk pemanggilan variasi, mula-mulai kita akan menghitung *read coverage* menggunakan bcftools. Selain, itu kita juga akan menandai posisi variasi (posisi sekuen sampel yang berbeda dengan sekuen referensi). Output dari proses penambahan kedua informasi tersebut adalah file dalam format BCF."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"a9ja38YTjGRb"},"outputs":[],"source":["#Sampel 1\n","! bcftools mpileup -O b -o 1_HBA2_raw.bcf -f data/ref_genome/chr16.fa results/bam/1_HBA2_sorted.bam #hitung read coverage dan tandai variasi\n","! mv *.bcf results/bcf #pindah output .bcf ke folder result"]},{"cell_type":"markdown","metadata":{"id":"NxLGZbxWjGRc"},"source":["Silahkan tulis command untuk sampel yang kedua, "]},{"cell_type":"code","execution_count":null,"metadata":{"id":"c2m9oxW5jGRc"},"outputs":[],"source":["#Sampel 2\n"]},{"cell_type":"markdown","metadata":{"id":"k7VX2ZMZjGRc"},"source":["## Deteksi polimorfisme nukleotida tunggal (SNP)"]},{"cell_type":"markdown","metadata":{"id":"1UvI2hhhjGRc"},"source":["Selanjutnya kita akan melakukan pemanggilan variasi yang sebelumnya sudah ditandai. Dengan menjalankan command dibawah kita akan memperoleh posisi dan jenis variasi beserta informasi lainnya, seperti *quality*."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"-sVIiA_xjGRc"},"outputs":[],"source":["#Sampel 1\n","! bcftools call --ploidy 1 -m -v -o 1_HBA2_variants.vcf results/bcf/1_HBA2_raw.bcf #pemanggilan variasi\n","! mv *.vcf results/vcf #pindah output .vcf ke folder result"]},{"cell_type":"markdown","metadata":{"id":"fHCn1j8hjGRd"},"source":["Silahkan tulis command untuk sampel yang kedua, "]},{"cell_type":"code","execution_count":null,"metadata":{"id":"-qtd0fEcjGRd"},"outputs":[],"source":["#Sampel 2\n"]},{"cell_type":"markdown","metadata":{"id":"C-fhsAizjGRh"},"source":["## Filter dan laporkan varian SNP dalam VCF (variant calling format) "]},{"cell_type":"markdown","metadata":{"id":"UTSRUle_jGRh"},"source":["Terakhir, kita melakukan filter untuk memperoleh variasi terakhir. "]},{"cell_type":"code","execution_count":null,"metadata":{"id":"EXzt8_O5jGRh"},"outputs":[],"source":["#Sampel 1\n","! vcfutils.pl varFilter results/vcf/1_HBA2_variants.vcf  > 1_HBA2_final_variants.vcf #pemanggilan variasi\n","! mv *.vcf results/vcf #pindah output .vcf ke folder result"]},{"cell_type":"markdown","metadata":{"id":"9GfCw5VtjGRh"},"source":["Silahkan tulis command untuk sampel yang kedua, "]},{"cell_type":"code","execution_count":null,"metadata":{"id":"PMIwG2VfjGRh"},"outputs":[],"source":["#Sampel 2\n"]},{"cell_type":"markdown","metadata":{"id":"eRw_d9tZjGRi"},"source":["### Lihat variasi pada VCF format"]},{"cell_type":"markdown","metadata":{"id":"FXIGx87hjGRi"},"source":["Silahkan run command berikut untuk melihat khusus pada bagian variasi yang terpanggil,"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Ba6NyL0gjGRi"},"outputs":[],"source":["! cat results/vcf/2_HBA2_final_variants.vcf | grep -A10 '#CHROM' "]},{"cell_type":"markdown","metadata":{"id":"QhmRe0aejGRi"},"source":["Untuk memahami masing-masing header silahkan baca laman tepatnya pada [*Explore the VCF format*](https://datacarpentry.org/wrangling-genomics/04-variant_calling/index.html)"]},{"cell_type":"markdown","metadata":{"id":"hZgOF03tjGRi"},"source":["## Alur kerja pemrosesan data NGS dan pemanggilan variasi"]},{"cell_type":"markdown","metadata":{"id":"bbO1GN7gjGRi"},"source":["![variant_calling_workflow.png](attachment:b955db7f-35ef-4fe5-86dc-b6f8bd1e56ac.png)"]},{"cell_type":"markdown","metadata":{"id":"X9m9eahQjGRi"},"source":["## Post-test / Assignment"]},{"cell_type":"markdown","metadata":{"id":"bH7EGeygjGRj"},"source":["Harap mengumpulkan,\n","\n","1. Screenshot visualisasi pemetaan short-read sampel 2 di IGV\n","2. Lokasi dan jenis variasi akhir pada sampel 2, beserta nilai quality nya\n","3. Screenshot visualisasi variasi sampel 1 dan sampel 2 di IGV, bagaimana variasi di kedua sampel? (bonus)\n","4. Bagaimana pengaruh variasi tersebut terhadap ekspresi gen HBA2? (bonus)"]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.3"},"colab":{"name":"Copy of Modules.ipynb","provenance":[]}},"nbformat":4,"nbformat_minor":0}