{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### [BIC30007 - _Sequence Alignment Fundamentals, Algorithm, and Application_](http://ugm.id/BIC30007practical02)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "### Pengaturan folder input dan output\n",
    "# Set up paths\n",
    "data_dir = Path('data')\n",
    "# Jika output dari data_dir.is_dir() adalah False, sesuaikan value dari data_dir dengan lokasi yang benar\n",
    "assert data_dir.is_dir(), f\"Folder {data_dir} tidak ditemukan, sesuaikan value dari data_dir dengan lokasi yang benar!\"\n",
    "\n",
    "# Set up output directory\n",
    "output_dir = Path('results/02.3')\n",
    "output_dir.mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Local Alignment\n",
    "#### Limitasi Algoritma Waterman\n",
    "Pada praktikum ini kita akan melakukan local alignment dari 2 sekuen, yaitu sequence1 dan sequence 2:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Waterman-Eggert local alignment of two sequences\n",
      "########################################\n",
      "# Program: matcher\n",
      "# Rundate: Sun  6 Mar 2022 21:11:37\n",
      "# Commandline: matcher\n",
      "#    [-asequence] data/sequence1.fasta\n",
      "#    [-bsequence] data/sequence2.fasta\n",
      "#    [-outfile] stdout\n",
      "# Align_format: markx0\n",
      "# Report_file: stdout\n",
      "########################################\n",
      "\n",
      "#=======================================\n",
      "#\n",
      "# Aligned_sequences: 2\n",
      "# 1: AJ002520.1\n",
      "# 2: AJ002520.1\n",
      "# Matrix: EDNAFULL\n",
      "# Gap_penalty: 16\n",
      "# Extend_penalty: 4\n",
      "#\n",
      "# Length: 191\n",
      "# Identity:     106/191 (55.5%)\n",
      "# Similarity:   106/191 (55.5%)\n",
      "# Gaps:           4/191 ( 2.1%)\n",
      "# Score: 142\n",
      "# \n",
      "#\n",
      "#=======================================\n",
      "\n",
      "           2080       2090      2100      2110      2120 \n",
      "AJ0025 CGCCACCAGG-ACACCTGCACGAGCCGAGCGGTGGAGGACCCGGGGCCCC\n",
      "       : :: ::::  ::  : :: :  ::: :::::::     :: :  :  ::\n",
      "AJ0025 CTCCCCCAGCTACTGCGGC-CTCGCCCAGCGGTGCTCCTCCAGCTGGTCC\n",
      "     390       400        410       420       430        \n",
      "\n",
      "            2130      2140      2150      2160      2170 \n",
      "AJ0025 AGGCGGTGGAGCTGGAGCGGCAGCAGCAGCGGCGCCCGGAGCCGGAGTGT\n",
      "       ::: ::::   ::::  : ::  : ::  :::     :::  :     : \n",
      "AJ0025 AGGTGGTGCTCCTGGTCCCGCTCCTGCCCCGGGCGTGGGATACACTCCGG\n",
      "      440       450       460       470       480        \n",
      "\n",
      "            2180      2190      2200      2210      2220 \n",
      "AJ0025 ATCCCACGCCCGGGGCAGGAGCGGGACCAGGAGCACCACCTGGACCAGCT\n",
      "        :::     :::  :: :  :: :  ::::   :::: :::::  :  : \n",
      "AJ0025 CTCCGGGCGCCGCTGCTGCTGCCGCTCCAGCTCCACCGCCTGGGGCCCCG\n",
      "      490       500       510       520       530        \n",
      "\n",
      "            2230      2240       2250      2260 \n",
      "AJ0025 GGAGGAGCACCGCTGGGCGAG-GCCGCAGTAGCTGGGGGAG\n",
      "       ::     ::::::: :::  : :: :  ::  :::: :: :\n",
      "AJ0025 GGTCCTCCACCGCTCGGCTCGTGCAGGTGTC-CTGGTGGCG\n",
      "      540       550       560        570        \n",
      "\n",
      "\n",
      "#---------------------------------------\n",
      "#---------------------------------------\n"
     ]
    }
   ],
   "source": [
    "! matcher {data_dir}/sequence1.fasta {data_dir}/sequence2.fasta stdout"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Seberapa mirip kah kedua sekuen tersebut?\n",
    "\n",
    "Mari kita coba _reverse_ salah satu sekuen di atas:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Waterman-Eggert local alignment of two sequences\n",
      "########################################\n",
      "# Program: matcher\n",
      "# Rundate: Sun  6 Mar 2022 21:11:37\n",
      "# Commandline: matcher\n",
      "#    -sreverse1\n",
      "#    [-asequence] data/sequence1.fasta\n",
      "#    [-bsequence] data/sequence2.fasta\n",
      "#    [-outfile] stdout\n",
      "# Align_format: markx0\n",
      "# Report_file: stdout\n",
      "########################################\n",
      "\n",
      "#=======================================\n",
      "#\n",
      "# Aligned_sequences: 2\n",
      "# 1: AJ002520.1\n",
      "# 2: AJ002520.1\n",
      "# Matrix: EDNAFULL\n",
      "# Gap_penalty: 16\n",
      "# Extend_penalty: 4\n",
      "#\n",
      "# Length: 2650\n",
      "# Identity:    2650/2650 (100.0%)\n",
      "# Similarity:  2650/2650 (100.0%)\n",
      "# Gaps:           0/2650 ( 0.0%)\n",
      "# Score: 13250\n",
      "# \n",
      "#\n",
      "#=======================================\n",
      "\n",
      "    2650      2640      2630      2620      2610         \n",
      "AJ0025 TTTTTTTTTTTTTTTTTTTACATATAATCGTACATTTAATTTATTTGCAT\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 TTTTTTTTTTTTTTTTTTTACATATAATCGTACATTTAATTTATTTGCAT\n",
      "               10        20        30        40        50\n",
      "\n",
      "    2600      2590      2580      2570      2560         \n",
      "AJ0025 ATTTAGCATACAGTCCAGAATGCGACGCACACTCAGTTATTAAAATATTA\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 ATTTAGCATACAGTCCAGAATGCGACGCACACTCAGTTATTAAAATATTA\n",
      "               60        70        80        90       100\n",
      "\n",
      "    2550      2540      2530      2520      2510         \n",
      "AJ0025 AAATTAGGCTTCGCGAAAAAAAAGTGATATAAACTTAATTTGAGAATGGG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 AAATTAGGCTTCGCGAAAAAAAAGTGATATAAACTTAATTTGAGAATGGG\n",
      "              110       120       130       140       150\n",
      "\n",
      "    2500      2490      2480      2470      2460         \n",
      "AJ0025 ATTGATGTGGAGTTCTTAGGTGTGGCGCCTAAATGGGGATTTCAAGTTAA\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 ATTGATGTGGAGTTCTTAGGTGTGGCGCCTAAATGGGGATTTCAAGTTAA\n",
      "              160       170       180       190       200\n",
      "\n",
      "    2450      2440      2430      2420      2410         \n",
      "AJ0025 GAGCATTCCGGTTGTTTGCCTTCCGGAAGTTCCGGGAGGGGACAGCTATG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 GAGCATTCCGGTTGTTTGCCTTCCGGAAGTTCCGGGAGGGGACAGCTATG\n",
      "              210       220       230       240       250\n",
      "\n",
      "    2400      2390      2380      2370      2360         \n",
      "AJ0025 TTACAACACCAGGGGATCGATTCATATATAAGTTAGGGACCACAACTAGG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 TTACAACACCAGGGGATCGATTCATATATAAGTTAGGGACCACAACTAGG\n",
      "              260       270       280       290       300\n",
      "\n",
      "    2350      2340      2330      2320      2310         \n",
      "AJ0025 ATCTGGGTGGTGGGATTCCTTGTCTTAGTCCGCCTCGTGCTTCTCCGCCT\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 ATCTGGGTGGTGGGATTCCTTGTCTTAGTCCGCCTCGTGCTTCTCCGCCT\n",
      "              310       320       330       340       350\n",
      "\n",
      "    2300      2290      2280      2270      2260         \n",
      "AJ0025 CTGGATGCTTGCCAGGTGCTGTTGCCGCTGGCGGAGCTACTCCCCCAGCT\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 CTGGATGCTTGCCAGGTGCTGTTGCCGCTGGCGGAGCTACTCCCCCAGCT\n",
      "              360       370       380       390       400\n",
      "\n",
      "    2250      2240      2230      2220      2210         \n",
      "AJ0025 ACTGCGGCCTCGCCCAGCGGTGCTCCTCCAGCTGGTCCAGGTGGTGCTCC\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 ACTGCGGCCTCGCCCAGCGGTGCTCCTCCAGCTGGTCCAGGTGGTGCTCC\n",
      "              410       420       430       440       450\n",
      "\n",
      "    2200      2190      2180      2170      2160         \n",
      "AJ0025 TGGTCCCGCTCCTGCCCCGGGCGTGGGATACACTCCGGCTCCGGGCGCCG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 TGGTCCCGCTCCTGCCCCGGGCGTGGGATACACTCCGGCTCCGGGCGCCG\n",
      "              460       470       480       490       500\n",
      "\n",
      "    2150      2140      2130      2120      2110         \n",
      "AJ0025 CTGCTGCTGCCGCTCCAGCTCCACCGCCTGGGGCCCCGGGTCCTCCACCG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 CTGCTGCTGCCGCTCCAGCTCCACCGCCTGGGGCCCCGGGTCCTCCACCG\n",
      "              510       520       530       540       550\n",
      "\n",
      "    2100      2090      2080      2070      2060         \n",
      "AJ0025 CTCGGCTCGTGCAGGTGTCCTGGTGGCGGTTGTTGTTGTTGCTGCTGCTG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 CTCGGCTCGTGCAGGTGTCCTGGTGGCGGTTGTTGTTGTTGCTGCTGCTG\n",
      "              560       570       580       590       600\n",
      "\n",
      "    2050      2040      2030      2020      2010         \n",
      "AJ0025 TGGTGGCGGTGGATTGCTTCCATGCTCCGGCGCATACGGATGCCCGGGAT\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 TGGTGGCGGTGGATTGCTTCCATGCTCCGGCGCATACGGATGCCCGGGAT\n",
      "              610       620       630       640       650\n",
      "\n",
      "    2000      1990      1980      1970      1960         \n",
      "AJ0025 TGCCATTGTGCTGCGGCGGCGGCGGACGCGGTGACATGTAGTGCGAGCCA\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 TGCCATTGTGCTGCGGCGGCGGCGGACGCGGTGACATGTAGTGCGAGCCA\n",
      "              660       670       680       690       700\n",
      "\n",
      "    1950      1940      1930      1920      1910         \n",
      "AJ0025 GGTGGCGGGCCATAGGACTGCTGATACGGCGAGAACGGTGCGTAGTAGGG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 GGTGGCGGGCCATAGGACTGCTGATACGGCGAGAACGGTGCGTAGTAGGG\n",
      "              710       720       730       740       750\n",
      "\n",
      "    1900      1890      1880      1870      1860         \n",
      "AJ0025 CTGTGGCGTGGGATGTCCGCCGTACTGGGGCTGGTAATAGGGTCGTGCTG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 CTGTGGCGTGGGATGTCCGCCGTACTGGGGCTGGTAATAGGGTCGTGCTG\n",
      "              760       770       780       790       800\n",
      "\n",
      "    1850      1840      1830      1820      1810         \n",
      "AJ0025 GTCCACCGGGTGGTCCAGGCGGCGGGCCATGTGGTGGTCCGCCGTAGTTA\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 GTCCACCGGGTGGTCCAGGCGGCGGGCCATGTGGTGGTCCGCCGTAGTTA\n",
      "              810       820       830       840       850\n",
      "\n",
      "    1800      1790      1780      1770      1760         \n",
      "AJ0025 GCATAACCCGTGTGCGGAGGTGGCGGCGGCAGACCGGGCATTCCCGGCGG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 GCATAACCCGTGTGCGGAGGTGGCGGCGGCAGACCGGGCATTCCCGGCGG\n",
      "              860       870       880       890       900\n",
      "\n",
      "    1750      1740      1730      1720      1710         \n",
      "AJ0025 CGGCTGATGCGGCCCCATGTGCGGTGGCATCAAGGAATGTCCGGGTGGCG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 CGGCTGATGCGGCCCCATGTGCGGTGGCATCAAGGAATGTCCGGGTGGCG\n",
      "              910       920       930       940       950\n",
      "\n",
      "    1700      1690      1680      1670      1660         \n",
      "AJ0025 GATGGCCGGGATGTGGATGCATCTGCGGCATGCCGGGCGGTGGAGCTGGC\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 GATGGCCGGGATGTGGATGCATCTGCGGCATGCCGGGCGGTGGAGCTGGC\n",
      "              960       970       980       990      1000\n",
      "\n",
      "    1650      1640      1630      1620      1610         \n",
      "AJ0025 CCGCCCTGTTGCTGCGGCAGTGGCACCTGTTGTGGCACCGGCGGTGGCGT\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 CCGCCCTGTTGCTGCGGCAGTGGCACCTGTTGTGGCACCGGCGGTGGCGT\n",
      "             1010      1020      1030      1040      1050\n",
      "\n",
      "    1600      1590      1580      1570      1560         \n",
      "AJ0025 TGGTGTCGTCGTCGTTGTTGTGGTCGCTGCAGTAGCTGCTCCGGCCGCTG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 TGGTGTCGTCGTCGTTGTTGTGGTCGCTGCAGTAGCTGCTCCGGCCGCTG\n",
      "             1060      1070      1080      1090      1100\n",
      "\n",
      "    1550      1540      1530      1520      1510         \n",
      "AJ0025 GAGCCACCTCAGGCGGCGGTGTGGGTGTCGGCACCGGTGTGGGCACCTTG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 GAGCCACCTCAGGCGGCGGTGTGGGTGTCGGCACCGGTGTGGGCACCTTG\n",
      "             1110      1120      1130      1140      1150\n",
      "\n",
      "    1500      1490      1480      1470      1460         \n",
      "AJ0025 CCGGGCTGGGCCACCACTTTGGCGGGTTCCGTTTTGGGTGGCTCGACTAG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 CCGGGCTGGGCCACCACTTTGGCGGGTTCCGTTTTGGGTGGCTCGACTAG\n",
      "             1160      1170      1180      1190      1200\n",
      "\n",
      "    1450      1440      1430      1420      1410         \n",
      "AJ0025 CGTTGACTGCGGAGGCAGGGCAGCGGCCATCTCCAGCTTCTCTGGCTTGA\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 CGTTGACTGCGGAGGCAGGGCAGCGGCCATCTCCAGCTTCTCTGGCTTGA\n",
      "             1210      1220      1230      1240      1250\n",
      "\n",
      "    1400      1390      1380      1370      1360         \n",
      "AJ0025 TGGGTGGCGGCGGAACCGATGGCTTCGGCGGTGGCTCAATGTCCATGGGC\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 TGGGTGGCGGCGGAACCGATGGCTTCGGCGGTGGCTCAATGTCCATGGGC\n",
      "             1260      1270      1280      1290      1300\n",
      "\n",
      "    1350      1340      1330      1320      1310         \n",
      "AJ0025 TCGGGATCCGTTTTGCTGGCAGTTTCGTGCACGTGCACCGCTGGCGCCGG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 TCGGGATCCGTTTTGCTGGCAGTTTCGTGCACGTGCACCGCTGGCGCCGG\n",
      "             1310      1320      1330      1340      1350\n",
      "\n",
      "    1300      1290      1280      1270      1260         \n",
      "AJ0025 CGTAGGCGTTGAACTCGGTGTGGGATTGAGTGTAGCGGGCTTCACAGCCG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 CGTAGGCGTTGAACTCGGTGTGGGATTGAGTGTAGCGGGCTTCACAGCCG\n",
      "             1360      1370      1380      1390      1400\n",
      "\n",
      "    1250      1240      1230      1220      1210         \n",
      "AJ0025 GCGGTGTCTCCTTGGGCGGAGCTGCTGGCGCTGGTACCGCCGACGCCGGC\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 GCGGTGTCTCCTTGGGCGGAGCTGCTGGCGCTGGTACCGCCGACGCCGGC\n",
      "             1410      1420      1430      1440      1450\n",
      "\n",
      "    1200      1190      1180      1170      1160         \n",
      "AJ0025 TCTTGGCCAGCGGGTGTGGCAGCGGGCTGACCTGGCTGCGTTGGCGGCTT\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 TCTTGGCCAGCGGGTGTGGCAGCGGGCTGACCTGGCTGCGTTGGCGGCTT\n",
      "             1460      1470      1480      1490      1500\n",
      "\n",
      "    1150      1140      1130      1120      1110         \n",
      "AJ0025 CACTGCCTCCTCGCTCCTCGTCACTGGAGCGGCGGCAGCTGTAGCGGCAG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 CACTGCCTCCTCGCTCCTCGTCACTGGAGCGGCGGCAGCTGTAGCGGCAG\n",
      "             1510      1520      1530      1540      1550\n",
      "\n",
      "    1100      1090      1080      1070      1060         \n",
      "AJ0025 CTGCAGCAGGATTCGCTGCGCTATTCGCATTCGCATTCGGTTCATCCAAT\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 CTGCAGCAGGATTCGCTGCGCTATTCGCATTCGCATTCGGTTCATCCAAT\n",
      "             1560      1570      1580      1590      1600\n",
      "\n",
      "    1050      1040      1030      1020      1010         \n",
      "AJ0025 CGCTGGCGATCGCGCTTGATGTCCTCGTACATGCGAAGCACCATCTTCTG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 CGCTGGCGATCGCGCTTGATGTCCTCGTACATGCGAAGCACCATCTTCTG\n",
      "             1610      1620      1630      1640      1650\n",
      "\n",
      "    1000       990       980       970       960         \n",
      "AJ0025 GAAAGTTTCCTCGTCGACCGCCGGCTTGCAGTGGCGCTTTAGCTCCTCCT\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 GAAAGTTTCCTCGTCGACCGCCGGCTTGCAGTGGCGCTTTAGCTCCTCCT\n",
      "             1660      1670      1680      1690      1700\n",
      "\n",
      "     950       940       930       920       910         \n",
      "AJ0025 GGAAGGCTTCGCTGGACTCCACCATGCGCTGCTTCTTGGCCTCGAACTTC\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 GGAAGGCTTCGCTGGACTCCACCATGCGCTGCTTCTTGGCCTCGAACTTC\n",
      "             1710      1720      1730      1740      1750\n",
      "\n",
      "     900       890       880       870       860         \n",
      "AJ0025 TCCTCCATCTGCTGCAGTTCCGCCTCAAGTTTGGTCTGGTGCATCGTCAA\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 TCCTCCATCTGCTGCAGTTCCGCCTCAAGTTTGGTCTGGTGCATCGTCAA\n",
      "             1760      1770      1780      1790      1800\n",
      "\n",
      "     850       840       830       820       810         \n",
      "AJ0025 CGAGCTGACCTGCCGCTTGAGCACCTGCATCCGCGTGGTGGTCACCACTG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 CGAGCTGACCTGCCGCTTGAGCACCTGCATCCGCGTGGTGGTCACCACTG\n",
      "             1810      1820      1830      1840      1850\n",
      "\n",
      "     800       790       780       770       760         \n",
      "AJ0025 AACGCACATCTGGTACAACGGCCTCCGAGAATATCTCGTTAATCAGGCGA\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 AACGCACATCTGGTACAACGGCCTCCGAGAATATCTCGTTAATCAGGCGA\n",
      "             1860      1870      1880      1890      1900\n",
      "\n",
      "     750       740       730       720       710         \n",
      "AJ0025 TGGTTGTGCAGATAGCGGGCGTAGGCCAAGTGCTTGGTCGTATACCCTTC\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 TGGTTGTGCAGATAGCGGGCGTAGGCCAAGTGCTTGGTCGTATACCCTTC\n",
      "             1910      1920      1930      1940      1950\n",
      "\n",
      "     700       690       680       670       660         \n",
      "AJ0025 GTCCTGGTCATCCTCGTCCTCCGCCGGCTGAATGTCGATGCGTCGCTCAT\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 GTCCTGGTCATCCTCGTCCTCCGCCGGCTGAATGTCGATGCGTCGCTCAT\n",
      "             1960      1970      1980      1990      2000\n",
      "\n",
      "     650       640       630       620       610         \n",
      "AJ0025 GCTGCGACTTACTGCCGCCACCGCGTGACGGTGTCTCATGCATATCCACG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 GCTGCGACTTACTGCCGCCACCGCGTGACGGTGTCTCATGCATATCCACG\n",
      "             2010      2020      2030      2040      2050\n",
      "\n",
      "     600       590       580       570       560         \n",
      "AJ0025 TCGGTCTTCACCTTGCTTTTGGCCGACATGTAGGCCTGATAGGCGGGCGT\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 TCGGTCTTCACCTTGCTTTTGGCCGACATGTAGGCCTGATAGGCGGGCGT\n",
      "             2060      2070      2080      2090      2100\n",
      "\n",
      "     550       540       530       520       510         \n",
      "AJ0025 CTGGTGGTAGGCCTTTAGCGATTTCTCGTACTCCAGCTTCTCCGCCTCGT\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 CTGGTGGTAGGCCTTTAGCGATTTCTCGTACTCCAGCTTCTCCGCCTCGT\n",
      "             2110      2120      2130      2140      2150\n",
      "\n",
      "     500       490       480       470       460         \n",
      "AJ0025 ATTCGTCAATGAACTCCGTCTTCTCATCCTCGGGCAGCAACTTCCACATG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 ATTCGTCAATGAACTCCGTCTTCTCATCCTCGGGCAGCAACTTCCACATG\n",
      "             2160      2170      2180      2190      2200\n",
      "\n",
      "     450       440       430       420       410         \n",
      "AJ0025 GCACCGATCTTCTTACCCAGCTCCCACAGCTTCAGCTCTGGATGCTTCGC\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 GCACCGATCTTCTTACCCAGCTCCCACAGCTTCAGCTCTGGATGCTTCGC\n",
      "             2210      2220      2230      2240      2250\n",
      "\n",
      "     400       390       380       370       360         \n",
      "AJ0025 CTTGACGCTATCCCAGACGCGTTTCGAATACCGCATGTAGGGCAAAATGG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 CTTGACGCTATCCCAGACGCGTTTCGAATACCGCATGTAGGGCAAAATGG\n",
      "             2260      2270      2280      2290      2300\n",
      "\n",
      "     350       340       330       320       310         \n",
      "AJ0025 GCTTCTCCGGCGGCTTTGGTGGCTTGGGTAGCCTTGATTCGTTCTGGTTC\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 GCTTCTCCGGCGGCTTTGGTGGCTTGGGTAGCCTTGATTCGTTCTGGTTC\n",
      "             2310      2320      2330      2340      2350\n",
      "\n",
      "     300       290       280       270       260         \n",
      "AJ0025 TTGCTGCTCGACGACTTGGTTACCTTCTGTGGCGTGAATGCCGGATTCCC\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 TTGCTGCTCGACGACTTGGTTACCTTCTGTGGCGTGAATGCCGGATTCCC\n",
      "             2360      2370      2380      2390      2400\n",
      "\n",
      "     250       240       230       220       210         \n",
      "AJ0025 ATAGTTGCTGTGCGTGAAGATGGGCGTCTGATCCTTGTTGCGATCGCCAC\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 ATAGTTGCTGTGCGTGAAGATGGGCGTCTGATCCTTGTTGCGATCGCCAC\n",
      "             2410      2420      2430      2440      2450\n",
      "\n",
      "     200       190       180       170       160         \n",
      "AJ0025 CACCTCCGCCGCCGGATGAACGCGACCGGCTGCCGCCGCCGGAGCCACCG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 CACCTCCGCCGCCGGATGAACGCGACCGGCTGCCGCCGCCGGAGCCACCG\n",
      "             2460      2470      2480      2490      2500\n",
      "\n",
      "     150       140       130       120       110         \n",
      "AJ0025 CCGCCGCCCTGGAGTGGTGTGGCCGATCCCTGGCCGCCAACGGCTATCTG\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 CCGCCGCCCTGGAGTGGTGTGGCCGATCCCTGGCCGCCAACGGCTATCTG\n",
      "             2510      2520      2530      2540      2550\n",
      "\n",
      "     100        90        80        70        60         \n",
      "AJ0025 CTTGTAGTTGCTTGGCAGGGCCATCTTTGTGCAATTCGCTGGTCAAATTT\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 CTTGTAGTTGCTTGGCAGGGCCATCTTTGTGCAATTCGCTGGTCAAATTT\n",
      "             2560      2570      2580      2590      2600\n",
      "\n",
      "      50        40        30        20        10         \n",
      "AJ0025 GTATGGATTTTCCAAGAGAAATAACAATTTTCAGACAGACGCCATTTTCT\n",
      "       ::::::::::::::::::::::::::::::::::::::::::::::::::\n",
      "AJ0025 GTATGGATTTTCCAAGAGAAATAACAATTTTCAGACAGACGCCATTTTCT\n",
      "             2610      2620      2630      2640      2650\n",
      "\n",
      "\n",
      "#---------------------------------------\n",
      "#---------------------------------------\n"
     ]
    }
   ],
   "source": [
    "! matcher -sreverse1 {data_dir}/sequence1.fasta {data_dir}/sequence2.fasta stdout"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Seberapa mirip kah kedua sekuen tersebut? Apa yang membedakan dengan alignment sebelumnya?\n",
    "* Apa yang bisa kamu simpulkan dari kedua sekuen tersebut?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Protein Homolog\n",
    "Mari kita coba sejajarkan sekuen DNA pengkode protein cyclin di manusia dan anjing:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Waterman-Eggert local alignment of two sequences\n",
      "########################################\n",
      "# Program: matcher\n",
      "# Rundate: Sun  6 Mar 2022 21:11:37\n",
      "# Commandline: matcher\n",
      "#    [-asequence] data/CDKL4.human.fasta\n",
      "#    [-bsequence] data/cyclin.dog.fasta\n",
      "#    [-outfile] stdout\n",
      "# Align_format: markx0\n",
      "# Report_file: stdout\n",
      "########################################\n",
      "\n",
      "#=======================================\n",
      "#\n",
      "# Aligned_sequences: 2\n",
      "# 1: CDKL4_HUMAN\n",
      "# 2: cyclin.dog\n",
      "# Matrix: EDNAFULL\n",
      "# Gap_penalty: 16\n",
      "# Extend_penalty: 4\n",
      "#\n",
      "# Length: 927\n",
      "# Identity:     859/927 (92.7%)\n",
      "# Similarity:   859/927 (92.7%)\n",
      "# Gaps:           0/927 ( 0.0%)\n",
      "# Score: 4023\n",
      "# \n",
      "#\n",
      "#=======================================\n",
      "\n",
      "               10        20        30        40        50\n",
      "CDKL4_ ATGGAAAAGTATGAAAAATTAGCTAAGACTGGAGAAGGGTCTTATGGGGT\n",
      "       ::::::::::: :::::::::::::::: :::::::::::::::::::::\n",
      "cyclin ATGGAAAAGTACGAAAAATTAGCTAAGATTGGAGAAGGGTCTTATGGGGT\n",
      "               10        20        30        40        50\n",
      "\n",
      "               60        70        80        90       100\n",
      "CDKL4_ TGTATTCAAATGCAGAAACAAAACCTCTGGACAAGTAGTAGCTGTTAAAA\n",
      "       ::::::::::::::::::::::::::: :::::::: :::::: ::::::\n",
      "cyclin TGTATTCAAATGCAGAAACAAAACCTCAGGACAAGTGGTAGCTATTAAAA\n",
      "               60        70        80        90       100\n",
      "\n",
      "              110       120       130       140       150\n",
      "CDKL4_ AATTTGTGGAATCTGAAGATGATCCTGTTGTTAAGAAAATAGCACTAAGA\n",
      "       ::::::::::::::::::::::::::::: ::::::::::::::::::: \n",
      "cyclin AATTTGTGGAATCTGAAGATGATCCTGTTATTAAGAAAATAGCACTAAGG\n",
      "              110       120       130       140       150\n",
      "\n",
      "              160       170       180       190       200\n",
      "CDKL4_ GAAATACGTATGTTGAAGCAATTAAAACATCCAAATCTTGTGAACCTCAT\n",
      "       ::::::::::::::::::::::::::::: ::::: ::::::::::::::\n",
      "cyclin GAAATACGTATGTTGAAGCAATTAAAACACCCAAACCTTGTGAACCTCAT\n",
      "              160       170       180       190       200\n",
      "\n",
      "              210       220       230       240       250\n",
      "CDKL4_ CGAGGTGTTCAGGAGAAAAAGGAAAATGCATTTAGTTTTTGAATACTGTG\n",
      "        ::::: ::::: ::::::::::::::::::::::: :::::::::::::\n",
      "cyclin TGAGGTTTTCAGAAGAAAAAGGAAAATGCATTTAGTATTTGAATACTGTG\n",
      "              210       220       230       240       250\n",
      "\n",
      "              260       270       280       290       300\n",
      "CDKL4_ ATCATACACTTTTAAATGAGCTGGAAAGAAACCCAAATGGAGTTGCTGAT\n",
      "       :::: :::::::::::::::::::::::::: ::::: ::::::::::::\n",
      "cyclin ATCACACACTTTTAAATGAGCTGGAAAGAAATCCAAAAGGAGTTGCTGAT\n",
      "              260       270       280       290       300\n",
      "\n",
      "              310       320       330       340       350\n",
      "CDKL4_ GGAGTGATCAAAAGCGTATTATGGCAAACACTTCAAGCTCTTAATTTCTG\n",
      "       :::::::::::::: :::::::::::::::::::::::::::::::::::\n",
      "cyclin GGAGTGATCAAAAGTGTATTATGGCAAACACTTCAAGCTCTTAATTTCTG\n",
      "              310       320       330       340       350\n",
      "\n",
      "              360       370       380       390       400\n",
      "CDKL4_ TCATATACATAACTGTATTCACAGAGATATAAAACCTGAAAATATTCTAA\n",
      "       ::::: :::::::::::::::::::::: :::::::::::::::::::::\n",
      "cyclin TCATAAACATAACTGTATTCACAGAGATGTAAAACCTGAAAATATTCTAA\n",
      "              360       370       380       390       400\n",
      "\n",
      "              410       420       430       440       450\n",
      "CDKL4_ TAACTAAGCAAGGAATAATCAAGATTTGTGACTTCGGGTTTGCACAAATT\n",
      "       ::::::::::::: ::::::::::: :::::::: :::::::::: ::::\n",
      "cyclin TAACTAAGCAAGGGATAATCAAGATCTGTGACTTTGGGTTTGCACGAATT\n",
      "              410       420       430       440       450\n",
      "\n",
      "              460       470       480       490       500\n",
      "CDKL4_ CTGATTCCAGGAGATGCCTACACCGATTATGTAGCTACGAGATGGTACCG\n",
      "       :::::::::::::: :::::::: :: ::::: :::::  ::::::::::\n",
      "cyclin CTGATTCCAGGAGACGCCTACACGGACTATGTCGCTACCCGATGGTACCG\n",
      "              460       470       480       490       500\n",
      "\n",
      "              510       520       530       540       550\n",
      "CDKL4_ AGCTCCTGAACTTCTTGTGGGAGATACTCAGTATGGTTCTTCAGTCGATA\n",
      "       :::::::::::::::::::::::: :: ::::::::::::::::: :: :\n",
      "cyclin AGCTCCTGAACTTCTTGTGGGAGACACACAGTATGGTTCTTCAGTAGACA\n",
      "              510       520       530       540       550\n",
      "\n",
      "              560       570       580       590       600\n",
      "CDKL4_ TATGGGCTATTGGTTGTGTTTTTGCAGAGCTCCTGACAGGCCAGCCACTG\n",
      "       ::::::::::::: :: :::::::::::::: : ::::::::::::::: \n",
      "cyclin TATGGGCTATTGGCTGCGTTTTTGCAGAGCTTCAGACAGGCCAGCCACTC\n",
      "              560       570       580       590       600\n",
      "\n",
      "              610       620       630       640       650\n",
      "CDKL4_ TGGCCTGGAAAATCAGATGTGGACCAACTTTATCTGATAATCAGAACACT\n",
      "       :::::::::::::::::::::::::: ::::::::::: :::::::::::\n",
      "cyclin TGGCCTGGAAAATCAGATGTGGACCAGCTTTATCTGATCATCAGAACACT\n",
      "              610       620       630       640       650\n",
      "\n",
      "              660       670       680       690       700\n",
      "CDKL4_ AGGAAAATTAATCCCAAGACATCAATCAATCTTTAAAAGTAACGGGTTTT\n",
      "        :::::: ::::::::::::::::::: ::::: :::::::::  :::::\n",
      "cyclin GGGAAAACTAATCCCAAGACATCAATCTATCTTCAAAAGTAACCAGTTTT\n",
      "              660       670       680       690       700\n",
      "\n",
      "              710       720       730       740       750\n",
      "CDKL4_ TCCATGGCATCAGTATACCTGAGCCAGAAGACATGGAAACTCTTGAGGAA\n",
      "       :::::::::::::::: ::::::::::::::::::::::::::::: :::\n",
      "cyclin TCCATGGCATCAGTATTCCTGAGCCAGAAGACATGGAAACTCTTGAAGAA\n",
      "              710       720       730       740       750\n",
      "\n",
      "              760       770       780       790       800\n",
      "CDKL4_ AAGTTCTCAGATGTTCATCCTGTGGCTCTGAACTTCATGAAGGGGTGTCT\n",
      "       ::::::::::::: ::: ::  :: :: :::: :::::::::: ::::::\n",
      "cyclin AAGTTCTCAGATGCTCACCCCATGTCTTTGAATTTCATGAAGGAGTGTCT\n",
      "              760       770       780       790       800\n",
      "\n",
      "              810       820       830       840       850\n",
      "CDKL4_ GAAGATGAATCCAGATGACAGATTAACCTGTTCCCAACTCCTGGAGAGCT\n",
      "       :::::::::::: :::::::::::::::::: ::::::::::::::::  \n",
      "cyclin GAAGATGAATCCTGATGACAGATTAACCTGTGCCCAACTCCTGGAGAGTC\n",
      "              810       820       830       840       850\n",
      "\n",
      "              860       870       880       890       900\n",
      "CDKL4_ CCTACTTTGATTCTTTTCAAGAGGCCCAAATTAAAAGAAAAGCACGTAAT\n",
      "       :::::::::::::::: :: :::: ::: :::::::::::::::: :: :\n",
      "cyclin CCTACTTTGATTCTTTCCACGAGGACCATATTAAAAGAAAAGCACATACT\n",
      "              860       870       880       890       900\n",
      "\n",
      "              910       920       \n",
      "CDKL4_ GAAGGAAGAAACAGAAGACGCCAACAG\n",
      "       :::::::::::::::::::: :: :::\n",
      "cyclin GAAGGAAGAAACAGAAGACGTCAGCAG\n",
      "              910       920       \n",
      "\n",
      "\n",
      "#---------------------------------------\n",
      "#---------------------------------------\n"
     ]
    }
   ],
   "source": [
    "! matcher {data_dir}/CDKL4.human.fasta {data_dir}/cyclin.dog.fasta stdout"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Seberapa mirip kah kedua sekuen tersebut?\n",
    "* Apakah kita bisa menggunakan algoritma global untuk melakukan pensejajaran pada kedua sekuen tersebut?\n",
    "\n",
    "#### Conserved regions\n",
    "Tidak semua _region_ dari dua protein homolog memiliki variasi yang mirip. Ada bagian dari protein yang lebih _conserve_ dibandingkan bagian yang lain.\n",
    "\n",
    "Mari kita coba sejajarkan sekuen frataxin dari manusia dan nyamuk Aedes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Waterman-Eggert local alignment of two sequences\n",
      "########################################\n",
      "# Program: matcher\n",
      "# Rundate: Sun  6 Mar 2022 21:11:37\n",
      "# Commandline: matcher\n",
      "#    [-asequence] data/frataxin.human.gb\n",
      "#    [-bsequence] data/frataxin.aedes.gb\n",
      "#    [-outfile] stdout\n",
      "# Align_format: markx0\n",
      "# Report_file: stdout\n",
      "########################################\n",
      "\n",
      "#=======================================\n",
      "#\n",
      "# Aligned_sequences: 2\n",
      "# 1: NP_852090\n",
      "# 2: XP_001656178\n",
      "# Matrix: EBLOSUM62\n",
      "# Gap_penalty: 14\n",
      "# Extend_penalty: 4\n",
      "#\n",
      "# Length: 70\n",
      "# Identity:      34/70 (48.6%)\n",
      "# Similarity:    50/70 (71.4%)\n",
      "# Gaps:           0/70 ( 0.0%)\n",
      "# Score: 196\n",
      "# \n",
      "#\n",
      "#=======================================\n",
      "\n",
      "      90       100       110       120       130         \n",
      "NP_852 LDETTYERLAEETLDSLAEFFEDLADKPYTFEDYDVSFGSGVLTVKLGGD\n",
      "       .:  :.: .  .::.:: ..:: : ..    .  :...: :::::  :  \n",
      "XP_001 IDSITFEAVCSDTLESLCDYFEQLVEEAAFLKSADITYGDGVLTVNFGEP\n",
      "         90       100       110       120       130      \n",
      "\n",
      "     140       150         \n",
      "NP_852 LGTYVINKQTPNKQIWLSSP\n",
      "        ::::::.:.::.:::::::\n",
      "XP_001 YGTYVINRQSPNRQIWLSSP\n",
      "        140       150      \n",
      "\n",
      "\n",
      "#---------------------------------------\n",
      "#---------------------------------------\n"
     ]
    }
   ],
   "source": [
    "! matcher {data_dir}/frataxin.human.gb {data_dir}/frataxin.aedes.gb stdout"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Mengapa ada _region_ yang tidak disejajarkan?\n",
    "* Mengapa suatu _region_ bisa lebih mirip dibandingkan _region_ yang lain?\n",
    "* Apakah ada yang spesial dari asam amino ke- 92, 100, 101, 104, 108, 111, 112, 122 and 124 pada protein frataxin di manusia? _**#hint :** buka file .gb dengan text editor._\n",
    "\n",
    "Mari kita coba memodifikasi parameter dari algoritma pensejajaran dengan mengubah gapopen set menjadi 4:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Waterman-Eggert local alignment of two sequences\n",
      "########################################\n",
      "# Program: matcher\n",
      "# Rundate: Sun  6 Mar 2022 21:11:38\n",
      "# Commandline: matcher\n",
      "#    [-asequence] data/frataxin.human.gb\n",
      "#    [-bsequence] data/frataxin.aedes.gb\n",
      "#    [-outfile] stdout\n",
      "#    -gapopen 4\n",
      "# Align_format: markx0\n",
      "# Report_file: stdout\n",
      "########################################\n",
      "\n",
      "#=======================================\n",
      "#\n",
      "# Aligned_sequences: 2\n",
      "# 1: NP_852090\n",
      "# 2: XP_001656178\n",
      "# Matrix: EBLOSUM62\n",
      "# Gap_penalty: 4\n",
      "# Extend_penalty: 4\n",
      "#\n",
      "# Length: 151\n",
      "# Identity:      59/151 (39.1%)\n",
      "# Similarity:    84/151 (55.6%)\n",
      "# Gaps:          24/151 (15.9%)\n",
      "# Score: 218\n",
      "# \n",
      "#\n",
      "#=======================================\n",
      "\n",
      "         40           50        60         70          80\n",
      "NP_852 G-RRGL-R-TD-IDATCTPRRASSNQRGL-NQIWNVKKQ-S-VYLMNLRK\n",
      "       : :: : :    :  :: :   :..:  :   :  :... . ..  .:  \n",
      "XP_001 GVRR-LPRCVSVIHQTC-PE--SAHQP-LPGGI-TVERRVNPIFRRHLSS\n",
      "               40           50          60        70     \n",
      "\n",
      "                90        100       110       120        \n",
      "NP_852 SGTLGHPG-SL-DETTYERLAEETLDSLAEFFEDLADKPYTF-EDYDVSF\n",
      "                :: :  :.: .  .::.:: ..:: : ..   : .  :...\n",
      "XP_001 MNPNDFIAVSLIDSITFEAVCSDTLESLCDYFEQLVEEA-AFLKSADITY\n",
      "          80        90       100       110        120    \n",
      "\n",
      "       130       140       150        160             170\n",
      "NP_852 GSGVLTVKLGGDLGTYVINKQTPNKQIWLSSP-S---RY-VV-DLS-VMT\n",
      "       : :::::  :   ::::::.:.::.::::::: :   ::  . : : :  \n",
      "XP_001 GDGVLTVNFGEPYGTYVINRQSPNRQIWLSSPKSGPKRYDFIPDKSKVNE\n",
      "          130       140       150       160       170    \n",
      "\n",
      "        \n",
      "NP_852 G\n",
      "       :\n",
      "XP_001 G\n",
      "        \n",
      "\n",
      "\n",
      "#---------------------------------------\n",
      "#---------------------------------------\n"
     ]
    }
   ],
   "source": [
    "! matcher {data_dir}/frataxin.human.gb {data_dir}/frataxin.aedes.gb stdout -gapopen 4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Apakah hasil pensejajaran ini lebih baik dari sebelumnya?\n",
    "\n",
    "Seperti yang sudah kita coba pada bagian sebelum ini, algoritma pensejajaran dapat memberikan lebih dari 1 alternatif, namun secara _default_, hanya satu hasil yang diberikan.\n",
    "\n",
    "Kita dapat memunculkan 3 alternatif lain dari pensejajaran diatas dengan menambahkan parameter berikut:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Waterman-Eggert local alignment of two sequences\n",
      "########################################\n",
      "# Program: matcher\n",
      "# Rundate: Sun  6 Mar 2022 21:11:38\n",
      "# Commandline: matcher\n",
      "#    [-asequence] data/frataxin.human.gb\n",
      "#    [-bsequence] data/frataxin.aedes.gb\n",
      "#    [-outfile] stdout\n",
      "#    -alternatives 3\n",
      "# Align_format: markx0\n",
      "# Report_file: stdout\n",
      "########################################\n",
      "\n",
      "#=======================================\n",
      "#\n",
      "# Aligned_sequences: 2\n",
      "# 1: NP_852090\n",
      "# 2: XP_001656178\n",
      "# Matrix: EBLOSUM62\n",
      "# Gap_penalty: 14\n",
      "# Extend_penalty: 4\n",
      "#\n",
      "# Length: 70\n",
      "# Identity:      34/70 (48.6%)\n",
      "# Similarity:    50/70 (71.4%)\n",
      "# Gaps:           0/70 ( 0.0%)\n",
      "# Score: 196\n",
      "# \n",
      "#\n",
      "#=======================================\n",
      "\n",
      "      90       100       110       120       130         \n",
      "NP_852 LDETTYERLAEETLDSLAEFFEDLADKPYTFEDYDVSFGSGVLTVKLGGD\n",
      "       .:  :.: .  .::.:: ..:: : ..    .  :...: :::::  :  \n",
      "XP_001 IDSITFEAVCSDTLESLCDYFEQLVEEAAFLKSADITYGDGVLTVNFGEP\n",
      "         90       100       110       120       130      \n",
      "\n",
      "     140       150         \n",
      "NP_852 LGTYVINKQTPNKQIWLSSP\n",
      "        ::::::.:.::.:::::::\n",
      "XP_001 YGTYVINRQSPNRQIWLSSP\n",
      "        140       150      \n",
      "\n",
      "\n",
      "#=======================================\n",
      "#\n",
      "# Aligned_sequences: 2\n",
      "# 1: NP_852090\n",
      "# 2: XP_001656178\n",
      "# Matrix: EBLOSUM62\n",
      "# Gap_penalty: 14\n",
      "# Extend_penalty: 4\n",
      "#\n",
      "# Length: 14\n",
      "# Identity:       4/14 (28.6%)\n",
      "# Similarity:     8/14 (57.1%)\n",
      "# Gaps:           0/14 ( 0.0%)\n",
      "# Score: 26\n",
      "# \n",
      "#\n",
      "#=======================================\n",
      "\n",
      "            150      \n",
      "NP_852 YVINKQTPNKQIWL\n",
      "       .. .:   :.  ::\n",
      "XP_001 FIPDKSKVNEGFWL\n",
      "          170        \n",
      "\n",
      "\n",
      "#=======================================\n",
      "#\n",
      "# Aligned_sequences: 2\n",
      "# 1: NP_852090\n",
      "# 2: XP_001656178\n",
      "# Matrix: EBLOSUM62\n",
      "# Gap_penalty: 14\n",
      "# Extend_penalty: 4\n",
      "#\n",
      "# Length: 18\n",
      "# Identity:       6/18 (33.3%)\n",
      "# Similarity:    10/18 (55.6%)\n",
      "# Gaps:           0/18 ( 0.0%)\n",
      "# Score: 24\n",
      "# \n",
      "#\n",
      "#=======================================\n",
      "\n",
      "        50        60     \n",
      "NP_852 ATCTPRRASSNQRGLNQI\n",
      "       ::      :. :::. ..\n",
      "XP_001 ATTMSLLVSAFQRGVRRL\n",
      "       20        30      \n",
      "\n",
      "\n",
      "#---------------------------------------\n",
      "#---------------------------------------\n"
     ]
    }
   ],
   "source": [
    "! matcher {data_dir}/frataxin.human.gb {data_dir}/frataxin.aedes.gb stdout -alternatives 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Global alignments\n",
    "Mari kita coba lakukan pensejajaran ulang dari kedua sekuen terakhir dengan menggunakan needle:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "########################################\n",
      "# Program: needle\n",
      "# Rundate: Sun  6 Mar 2022 21:11:38\n",
      "# Commandline: needle\n",
      "#    [-asequence] data/frataxin.human.gb\n",
      "#    [-bsequence] data/frataxin.aedes.gb\n",
      "#    [-outfile] stdout\n",
      "#    -auto\n",
      "# Align_format: srspair\n",
      "# Report_file: stdout\n",
      "########################################\n",
      "\n",
      "#=======================================\n",
      "#\n",
      "# Aligned_sequences: 2\n",
      "# 1: NP_852090\n",
      "# 2: XP_001656178\n",
      "# Matrix: EBLOSUM62\n",
      "# Gap_penalty: 10.0\n",
      "# Extend_penalty: 0.5\n",
      "#\n",
      "# Length: 276\n",
      "# Identity:      56/276 (20.3%)\n",
      "# Similarity:    87/276 (31.5%)\n",
      "# Gaps:         141/276 (51.1%)\n",
      "# Score: 178.5\n",
      "# \n",
      "#\n",
      "#=======================================\n",
      "\n",
      "NP_852090          1 ------------mwtlgrravagllaspspaqaqtltrvprpaelaplcg     38\n",
      "                                 :::|.......||.|   |..:.:.|:||...:     \n",
      "XP_001656178       1 mkrqkikaaqkqlfslyhattmsllvs---afqrgvrrlprcvsv-----     42\n",
      "\n",
      "NP_852090         39 rrglrtdidatctprras-------snqrglnqiwnvkkqsvylmnlrks     81\n",
      "                            |..|| |..|.       :.:|.:|.|:   ::.:..||    \n",
      "XP_001656178      43 -------ihqtc-pesahqplpggitverrvnpif---rrhlssmn----     77\n",
      "\n",
      "NP_852090         82 gtlghpgs------ldettyerlaeetldslaeffedladkpytfedydv    125\n",
      "                          |..      :|..|:|.:..:||:||.::||.|.::....:..|:\n",
      "XP_001656178      78 -----pndfiavslidsitfeavcsdtleslcdyfeqlveeaaflksadi    122\n",
      "\n",
      "NP_852090        126 sfgsgvltvklggdlgtyvinkqtpnkqiwlsspsryvvdlsvmtglgkt    175\n",
      "                     ::|.|||||..|...||||||:|:||:|||||||              |:\n",
      "XP_001656178     123 tygdgvltvnfgepygtyvinrqspnrqiwlssp--------------ks    158\n",
      "\n",
      "NP_852090        176 gctpttacpsmscwpqsslkp-----------------------------    196\n",
      "                     |                   |                             \n",
      "XP_001656178     159 g-------------------pkrydfipdkskvnegfwlykhdgvslhel    189\n",
      "\n",
      "NP_852090        196 --------------------------    196\n",
      "                                               \n",
      "XP_001656178     190 lqkeigvivqkevdflslphskrldn    215\n",
      "\n",
      "\n",
      "#---------------------------------------\n",
      "#---------------------------------------\n"
     ]
    }
   ],
   "source": [
    "! needle {data_dir}/frataxin.human.gb {data_dir}/frataxin.aedes.gb stdout -auto"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Apakah hasilnya lebih baik atau lebih buruk dari local alignment? Kenapa?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aplikasi: Mencari Ekson & Intron\n",
    "Kita dapat menggunakan alignment untuk menentukan ekson dan intron dari suatu gen. Kita akan mencoba membandingkan bagian genom dan cDNA dari gen AT3G4905 Arabidopsis.\n",
    "\n",
    "EMBOSS memiliki program yang terspesialisasi untuk melakukan analisis mRNA dari sekuen genomik (Expressed Sequence Tags / EST)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note Best alignment is between forward est and forward genome, and splice sites imply forward gene\n",
      "Exon        91 100.0     1    91 AT3G52905.1.genome     1    91 AT3G52905.1.cdna  \n",
      "+Intron    -20   0.0    92   271 AT3G52905.1.genome\n",
      "Exon       152 100.0   272   423 AT3G52905.1.genome    92   243 AT3G52905.1.cdna  \n",
      "+Intron    -20   0.0   424   547 AT3G52905.1.genome\n",
      "Exon        58 100.0   548   605 AT3G52905.1.genome   244   301 AT3G52905.1.cdna  \n",
      "+Intron    -20   0.0   606  1019 AT3G52905.1.genome\n",
      "Exon        69 100.0  1020  1088 AT3G52905.1.genome   302   370 AT3G52905.1.cdna  \n",
      "+Intron    -20   0.0  1089  1180 AT3G52905.1.genome\n",
      "Exon        96 100.0  1181  1276 AT3G52905.1.genome   371   466 AT3G52905.1.cdna  \n",
      "+Intron    -20   0.0  1277  1370 AT3G52905.1.genome\n",
      "Exon        90 100.0  1371  1460 AT3G52905.1.genome   467   556 AT3G52905.1.cdna  \n",
      "+Intron    -20   0.0  1461  1553 AT3G52905.1.genome\n",
      "Exon       185 100.0  1554  1738 AT3G52905.1.genome   557   741 AT3G52905.1.cdna  \n",
      "\n",
      "Span       621 100.0     1  1738 AT3G52905.1.genome     1   741 AT3G52905.1.cdna  \n",
      "\n",
      "Segment     91 100.0     1    91 AT3G52905.1.genome     1    91 AT3G52905.1.cdna  \n",
      "Segment    152 100.0   272   423 AT3G52905.1.genome    92   243 AT3G52905.1.cdna  \n",
      "Segment     58 100.0   548   605 AT3G52905.1.genome   244   301 AT3G52905.1.cdna  \n",
      "Segment     69 100.0  1020  1088 AT3G52905.1.genome   302   370 AT3G52905.1.cdna  \n",
      "Segment     96 100.0  1181  1276 AT3G52905.1.genome   371   466 AT3G52905.1.cdna  \n",
      "Segment     90 100.0  1371  1460 AT3G52905.1.genome   467   556 AT3G52905.1.cdna  \n",
      "Segment    185 100.0  1554  1738 AT3G52905.1.genome   557   741 AT3G52905.1.cdna  \n"
     ]
    }
   ],
   "source": [
    "! est2genome {data_dir}/AT3G4905.1.cdna.fasta {data_dir}/AT3G4905.1.genom.fasta stdout -auto"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Apa ciri khas dari mRNA sehingga perlu dibuat algortima pensejajaran yang lebih spesifik?\n",
    "* Ada berapa intron yang terdapat pada AT3G4905?\n",
    "\n",
    "Kita  dapat memperoleh juga hasil pensejajaran dengan menambahkan parameter berikut:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note Best alignment is between forward est and forward genome, and splice sites imply forward gene\n",
      "Exon        91 100.0     1    91 AT3G52905.1.genome     1    91 AT3G52905.1.cdna  \n",
      "+Intron    -20   0.0    92   271 AT3G52905.1.genome\n",
      "Exon       152 100.0   272   423 AT3G52905.1.genome    92   243 AT3G52905.1.cdna  \n",
      "+Intron    -20   0.0   424   547 AT3G52905.1.genome\n",
      "Exon        58 100.0   548   605 AT3G52905.1.genome   244   301 AT3G52905.1.cdna  \n",
      "+Intron    -20   0.0   606  1019 AT3G52905.1.genome\n",
      "Exon        69 100.0  1020  1088 AT3G52905.1.genome   302   370 AT3G52905.1.cdna  \n",
      "+Intron    -20   0.0  1089  1180 AT3G52905.1.genome\n",
      "Exon        96 100.0  1181  1276 AT3G52905.1.genome   371   466 AT3G52905.1.cdna  \n",
      "+Intron    -20   0.0  1277  1370 AT3G52905.1.genome\n",
      "Exon        90 100.0  1371  1460 AT3G52905.1.genome   467   556 AT3G52905.1.cdna  \n",
      "+Intron    -20   0.0  1461  1553 AT3G52905.1.genome\n",
      "Exon       185 100.0  1554  1738 AT3G52905.1.genome   557   741 AT3G52905.1.cdna  \n",
      "\n",
      "Span       621 100.0     1  1738 AT3G52905.1.genome     1   741 AT3G52905.1.cdna  \n",
      "\n",
      "Segment     91 100.0     1    91 AT3G52905.1.genome     1    91 AT3G52905.1.cdna  \n",
      "Segment    152 100.0   272   423 AT3G52905.1.genome    92   243 AT3G52905.1.cdna  \n",
      "Segment     58 100.0   548   605 AT3G52905.1.genome   244   301 AT3G52905.1.cdna  \n",
      "Segment     69 100.0  1020  1088 AT3G52905.1.genome   302   370 AT3G52905.1.cdna  \n",
      "Segment     96 100.0  1181  1276 AT3G52905.1.genome   371   466 AT3G52905.1.cdna  \n",
      "Segment     90 100.0  1371  1460 AT3G52905.1.genome   467   556 AT3G52905.1.cdna  \n",
      "Segment    185 100.0  1554  1738 AT3G52905.1.genome   557   741 AT3G52905.1.cdna  \n",
      "\n",
      "\n",
      "AT3G52905.1.genome vs AT3G52905.1.cdna:\n",
      "\n",
      "AT3G52905.1.genome      1 TGCTTCAATCGCGAAGACATTTAGCAGAACCATTTGTCGTTTCCTCTCGC     50\n",
      "                          ||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "  AT3G52905.1.cdna      1 TGCTTCAATCGCGAAGACATTTAGCAGAACCATTTGTCGTTTCCTCTCGC     50\n",
      "\n",
      "AT3G52905.1.genome     51 CAGATTCCGTGCTCCGGAGCCAGAAAGTTTCCGCCTTTAAAgtacg....     91\n",
      "                          |||||||||||||||||||||||||||||||||||||||||>>>>> 180\n",
      "  AT3G52905.1.cdna     51 CAGATTCCGTGCTCCGGAGCCAGAAAGTTTCCGCCTTTAAA.........     91\n",
      "\n",
      "AT3G52905.1.genome     91 .atcagGAAATGAAGTATGTGAAGCCATTGAGCTTATTGGGTAATGCGCT    315\n",
      "                           >>>>>||||||||||||||||||||||||||||||||||||||||||||\n",
      "  AT3G52905.1.cdna     91 ......GAAATGAAGTATGTGAAGCCATTGAGCTTATTGGGTAATGCGCT    135\n",
      "\n",
      "AT3G52905.1.genome    316 GAAGACCAAAGTGTCAGTCCCTGGTCGGTTTCTGGGTTTAGATGTTGGTG    365\n",
      "                          ||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "  AT3G52905.1.cdna    136 GAAGACCAAAGTGTCAGTCCCTGGTCGGTTTCTGGGTTTAGATGTTGGTG    185\n",
      "\n",
      "AT3G52905.1.genome    366 ATAAGTATGTTGGATTAGCTATCTCAGATCCTTCAAATATGGTTGCTTCT    415\n",
      "                          ||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "  AT3G52905.1.cdna    186 ATAAGTATGTTGGATTAGCTATCTCAGATCCTTCAAATATGGTTGCTTCT    235\n",
      "\n",
      "AT3G52905.1.genome    416 CCATTGAGgtacc.....tgcagTGTTTTGCTCAGAAAGAAATCAAACAT    574\n",
      "                          ||||||||>>>>> 124 >>>>>|||||||||||||||||||||||||||\n",
      "  AT3G52905.1.cdna    236 CCATTGAG...............TGTTTTGCTCAGAAAGAAATCAAACAT    270\n",
      "\n",
      "AT3G52905.1.genome    575 TGACCTGATGGCTACAGATTTCCAGAACCTGgtagg.....tgcagGTCA   1023\n",
      "                          |||||||||||||||||||||||||||||||>>>>> 414 >>>>>||||\n",
      "  AT3G52905.1.cdna    271 TGACCTGATGGCTACAGATTTCCAGAACCTG...............GTCA    305\n",
      "\n",
      "AT3G52905.1.genome   1024 AAGCATTTTCTGTGTCGGGATTAGTCGTTGGTTATCCATTTGGCAAACTG   1073\n",
      "                          ||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "  AT3G52905.1.cdna    306 AAGCATTTTCTGTGTCGGGATTAGTCGTTGGTTATCCATTTGGCAAACTG    355\n",
      "\n",
      "AT3G52905.1.genome   1074 AACAATGTAGAGGATgtaag....aacagGTTGTCACTGTGAATCTTTTC   1201\n",
      "                          |||||||||||||||>>>>> 92 >>>>>|||||||||||||||||||||\n",
      "  AT3G52905.1.cdna    356 AACAATGTAGAGGAT..............GTTGTCACTGTGAATCTTTTC    391\n",
      "\n",
      "AT3G52905.1.genome   1202 ATTGAGGAACTTCGTAAGACCGAAAAACTCAAGGATGTGAAATACACATA   1251\n",
      "                          ||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "  AT3G52905.1.cdna    392 ATTGAGGAACTTCGTAAGACCGAAAAACTCAAGGATGTGAAATACACATA    441\n",
      "\n",
      "AT3G52905.1.genome   1252 TTGGGACGAGCGATTATCATCAAAGgtata....aacagACCGTTGAACT   1381\n",
      "                          |||||||||||||||||||||||||>>>>> 94 >>>>>|||||||||||\n",
      "  AT3G52905.1.cdna    442 TTGGGACGAGCGATTATCATCAAAG..............ACCGTTGAACT    477\n",
      "\n",
      "AT3G52905.1.genome   1382 GATGTTGAAGCCCTTGAATTTGCATCCTGTTCAAGAGAAGACAATGTTGG   1431\n",
      "                          ||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "  AT3G52905.1.cdna    478 GATGTTGAAGCCCTTGAATTTGCATCCTGTTCAAGAGAAGACAATGTTGG    527\n",
      "\n",
      "AT3G52905.1.genome   1432 ACAAGTTAGCCGCAGTAGTTATACTTCAGgtaag....tgcagGAGTATT   1560\n",
      "                          |||||||||||||||||||||||||||||>>>>> 93 >>>>>|||||||\n",
      "  AT3G52905.1.cdna    528 ACAAGTTAGCCGCAGTAGTTATACTTCAG..............GAGTATT    563\n",
      "\n",
      "AT3G52905.1.genome   1561 TAGATTACGCGAACAGGTATGTAAACACTGAGCCAGCAGAGTAAAAGGAA   1610\n",
      "                          ||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "  AT3G52905.1.cdna    564 TAGATTACGCGAACAGGTATGTAAACACTGAGCCAGCAGAGTAAAAGGAA    613\n",
      "\n",
      "AT3G52905.1.genome   1611 ACGGGTAATGACAAAATAGGGAAGCCGCTAGTTTCGGATATATGAATATA   1660\n",
      "                          ||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "  AT3G52905.1.cdna    614 ACGGGTAATGACAAAATAGGGAAGCCGCTAGTTTCGGATATATGAATATA    663\n",
      "\n",
      "AT3G52905.1.genome   1661 ATGATGCTTGATTTGTTTTGACATTATCTATGTGTACAAAGACATCTTGC   1710\n",
      "                          ||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "  AT3G52905.1.cdna    664 ATGATGCTTGATTTGTTTTGACATTATCTATGTGTACAAAGACATCTTGC    713\n",
      "\n",
      "AT3G52905.1.genome   1711 ATCTTTTAAGAAATTCGTTTTTCGCTGC   1738\n",
      "                          ||||||||||||||||||||||||||||\n",
      "  AT3G52905.1.cdna    714 ATCTTTTAAGAAATTCGTTTTTCGCTGC    741\n",
      "\n",
      "\n",
      "Alignment Score: 621\n"
     ]
    }
   ],
   "source": [
    "! est2genome {data_dir}/AT3G4905.1.cdna.fasta {data_dir}/AT3G4905.1.genom.fasta stdout -align -auto"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Apakah needle atau matcher menghasilkan output alignment yang sama?\n",
    "\n",
    "## <mark>Aktivitas 1<mark>\n",
    "* <mark>Coba lakukan pensejajaran antara CDS dengan sekuen genom! Apakah hasilnya mirip dengan hasil pensejajaran dengan cDNA?<mark>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your code here\n",
    "# hint: gunakan est2genome\n",
    "# hint: file CDS ada di Data/AT3G4905.1.cds.fasta\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* <mark>Coba lakukan pensejajaran antara CDS dengan mRNA! Apakah yang dapat anda simpulkan?<mark>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your code here\n",
    "# hint: gunakan matcher / needle\n",
    "# hint: file CDS ada di Data/AT3G4905.1.cds.fasta\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <mark>Aktivitas 2<mark>\n",
    "<mark>Anda hendak melakukan amplifikasi sebagian sekuen dari Cyclin CDS dengan menggunakan dua primer: (1) primerf dan (2) primerr.<mark>\n",
    "\n",
    "* <mark>Coba cari lokasi dari kedua primer tersebut!<mark>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your code here\n",
    "# hint: gunakan matcher\n",
    "# hint: sekuen yang digunakan yaitu Data/CDKL4.human.fasta\n",
    "# hint: primer f ada di Data/primerf.fasta\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your code here\n",
    "# Write your code here\n",
    "# hint: gunakan matcher\n",
    "# hint: sekuen yang digunakan yaitu Data/CDKL4.human.fasta\n",
    "# hint: primer f ada di Data/primerr.fasta\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
